### PR TITLE
[BUGFIX] Removing measurement process hash/equality warnings

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Updating pennylane to stop relying on operator and measurement process hash and equality to get rid of warnings.

**Description of the Change:**
* Update `MeasurementValue._merge` to filter warnings raised by `MeasurementProcess` `__eq__` and `__hash__`. I chose to filter the warnings instead of changing the implementation because the new hash and equality will still be valid.
* Added dispatch for `qml.equal` to handle `MidMeasureMP`. Checking wires is not enough, so we need slightly more specific checks for two `MidMeasureMP`s to be equivalent.
* Updated `sum_expand` to stop using `list_of_measurements.index`.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/pauli/utils.py [hunk 1]
```diff
@@ -1366,11 +1366,15 @@ def simplify(h, cutoff=1.0e-12):
     for i, op in enumerate(h.ops):
         op = qml.operation.Tensor(op).prune()
         op = qml.pauli.pauli_word_to_string(op, wire_map=wiremap)
-        if op not in o:
```
Comment (timmysilv) [inline pennylane/pauli/utils.py:1369]: I think `op` is a string, and `o` is a list of strings here

File: tests/ops/functions/test_equal.py [hunk 2]
```diff
@@ -1236,6 +1236,26 @@ def test_observables_equal_but_wire_order_not(self):
...
+    def test_mid_measure(self):
+        """Test that `MidMeasureMP`s are equal only if their wires
+        an id are equal and their `reset` attribute match."""
+        mp = qml.measurements.MidMeasureMP(wires=qml.wires.Wires([0, 1]), reset=True, id="test_id")
+
+        mp1 = qml.measurements.MidMeasureMP(wires=qml.wires.Wires([1, 0]), reset=True, id="test_id")
+        mp2 = qml.measurements.MidMeasureMP(
+            wires=qml.wires.Wires([0, 1]), reset=False, id="test_id"
+        )
+        mp3 = qml.measurements.MidMeasureMP(wires=qml.wires.Wires([1, 0]), reset=True, id="foo")
```
Comment (timmysilv) [inline tests/ops/functions/test_equal.py:1248]:
```suggestion
        mp3 = qml.measurements.MidMeasureMP(wires=qml.wires.Wires([0, 1]), reset=True, id="foo")
```

**Review by lillian542 — APPROVED** [review]
Looks good to me 🚀

**Review by timmysilv — APPROVED** [review]
I am scared to approve bc I'm sure we're missing some spots... but what the heck right

**Pull request merged**